File: python/list/LC1.py
nums = [2, 7, 11, 15]
target = 9


class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        l = len(nums)

        for i in range(l-1):
            for j in range(i+1, l):
                if nums[i] + nums[j] == target:
                    return [nums[i], nums[j]]

    def twoSum(self, nums: list[int], target: int) -> list[int]:
        for i, n in enumerate(nums):
            complement = target - n

            if complement in nums[i+1:]:
                return [nums.index(n), nums[i+1:].index(complement) + (i + 1)]

    def twoSum(self, nums: list[int], target: int) -> list[int]:
        nums_map = {}
        for i, num in enumerate(nums):
            nums_map[num] = i

        for i, num in enumerate(nums):
            if target - num in nums_map and i !=nums_map[target-num]:
                return [i, nums_map[target - num]]

    def twoSum(self, nums: list[int], target: int) -> list[int]:
        nums_map = {}
        for i, num in enumerate(nums):
            if target - num in nums_map:
                return [nums_map[target-num], i]
            nums_map[num] = i

File: solution/src/leetcode/LC1.java
package leetcode;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class LC1 {
	class Solution {
		public int[] twoSum(int[] nums, int target) {
			for (int i = 0; i < nums.length - 1; i++) {
				for (int j = i + 1; j < nums.length; j++) {
					if (nums[i] + nums[j] == target) {
						return new int[] {i, j};
					}
				}
			}
			return new int[] {};
		}
	}

	class Solution2 {
		public int[] twoSum(int[] nums, int target) {
			int[] rst = new int[2];
			Elem[] elems = new Elem[nums.length];
			for (int i = 0; i < nums.length; i++) {
				elems[i] = new Elem(i, nums[i]);
			}
			Arrays.sort(elems, (s1, s2) -> {
				return s1.value - s2.value;
			});

			int start = 0;
			int end = nums.length - 1;
			while (start < end) {
				int sum = elems[start].value + elems[end].value;
				if (sum == target) {
					rst[0] = elems[start].index;
					rst[1] = elems[end].index;
					break;
				} else if (sum < target) {
					start++;
				} else {
					end--;
				}
			}
			return rst;
		}
	}

	class Elem {
		int index;
		int value;

		public Elem(int index, int value) {
			this.index = index;
			this.value = value;
		}
	}

	class Solution3 {
		public int[] twoSum(int[] nums, int target) {
			Map<Integer, Integer> numsMap = new HashMap<>();
			for (int i = 0; i < nums.length; i++) {
				if (numsMap.containsKey(target - nums[i])) {
					return new int[] {i, numsMap.get(target - nums[i])};
				}
				numsMap.put(nums[i], i);
			}
			return null;
		}
	}
}
